File: 1492-time-needed-to-inform-all-employees/time-needed-to-inform-all-employees.cpp
class Solution {
public:
    int bfs(int headID, vector<int>& informTime, vector<vector<int>> &adj)
    {
        queue<pair<int,int>> q;
        q.push({headID, 0});

        int minTime = 0;
        while(!q.empty())
        {
            auto p = q.front();
            q.pop();

            int manager = p.first;
            int time = p.second;

            minTime = max(minTime, time);

            for(int subOrdinates : adj[manager])
            q.push({subOrdinates, time + informTime[manager]});
    
        }

        return minTime;
    }

    int numOfMinutes(int n, int headID, vector<int>& manager, vector<int>& informTime) 
    {
        vector<vector<int>> adj(n);
        for(int i = 0; i < manager.size(); i++)
        {
            if(i != headID)
            adj[manager[i]].push_back(i);
        }

        return bfs(headID, informTime, adj);
    }
};

File: 1492-time-needed-to-inform-all-employees/time-needed-to-inform-all-employees.java
class Solution {
    private int timeToInformAll(int id, Map<Integer, List<Integer>> heirarchyMap, int[] informTime){
        if(!heirarchyMap.containsKey(id)){
            return 0;
        }

        int maxTime = 0;
        List<Integer> subordinates = heirarchyMap.get(id);
        for(int i = 0; i < subordinates.size(); i++){
            maxTime = Math.max(maxTime, timeToInformAll(subordinates.get(i), heirarchyMap, informTime));
        }
        return maxTime + informTime[id];
    }

    public int numOfMinutes(int n, int headID, int[] manager, int[] informTime) {
        Map<Integer, List<Integer>> heirarchyMap = new HashMap<>();
        for(int i = 0; i < n; i++) {
            heirarchyMap.putIfAbsent(manager[i], new ArrayList<>());
            heirarchyMap.get(manager[i]).add(i);
        }
        return timeToInformAll(headID, heirarchyMap, informTime);
    }
}